src/liballoc/collections/vec_deque.rs

```diff
@@ -1834,8 +1834,144 @@ impl<T> VecDeque<T> {
     #[inline]
     #[stable(feature = "append", since = "1.4.0")]
     pub fn append(&mut self, other: &mut Self) {
-        // naive impl
-        self.extend(other.drain(..));
+        // Copy from src[i1..i1 + len] to dst[i2..i2 + len].
+        // Does not check if the ranges are valid.
+        unsafe fn copy_part<T>(i1: usize, i2: usize, len: usize, src: &[T], dst: &mut [T]) {
```

Comment:
I'm missing why this part needs to be unsafe. Can't we build slices and use `[T]::copy_from_slice`?

Reply:
Ah, because `T` doesn't necessarily implement `Copy`, of course. However, it feels odd to "reimplement" a half-slice on top of a slice due to the `i1` / `i2`.

`i1` and `i2`  are actually... offsets(?) into `src` and `dst` respectively? Those should be renamed to at least have `src` and `dst` in there.